enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+replaces:
+superseded-by:
+---
+
+# Configurable DNS Forwarding
+
+This proposal allows cluster operators the ability to configure DNS name query forwarding.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+1. Is `operator.openshift.io` the best API group for the proposed `Upstream` type?
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+
+## Summary
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
+
+## Motivation
+
+DNS name resolution for pods is currently provided by an instance of [CoreDNS](https://coredns.io) that runs on each
+node in the cluster. If a DNS query is for a hostname within the cluster domain suffix (i.e. `cluster.local`), CoreDNS
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
+2. A minimal API surface that can be expanded to support future DNS forwarding use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+
+## Proposal
+
+The proposed `Upstream` type defines a schema for associating one or more `Nameservers` to a DNS subdomain identified by
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
+	Nameservers []string `json:"nameservers"`
+}
+```
+
+Each instance of CoreDNS performs health checking of `Nameservers`. If `Nameservers` consists of more than one
+nameserver, forwarding will be randomized on first use. When a healthy nameserver returns an error during the exchange,
```

Comment:
Users may be populating the forwarding list with entries they would usually put in the node `/etc/resolv.conf`. In these cases, could users reasonably expect the `resolv.conf` query-ordering semantics to apply to the custom nameservers list? If so, randomization may be seen as a bug (https://github.com/openshift/cluster-dns-operator/pull/129). A potential follow-on implication might be that forwarding policy configuration API is required, or a change in default to a sequential policy.

Reply:
@ironcladlou to address the issue you raise, I've added the `UpstreamPolicy` type to the proposed API , PTAL.